libusb/os/windows_hotplug.c

```diff
@@ -0,0 +1,296 @@
+#include "libusbi.h"
+#include "threads_windows.h"
+
+#include "windows_common.h"
+#include "windows_hotplug.h"
+
+#include <stdio.h>
+#include <dbt.h>
+#include <usbiodef.h>
+
+static HWND windows_event_hWnd;
+static HANDLE windows_event_thread_handle;
+static DWORD WINAPI windows_event_thread_main(LPVOID lpParam);
+static LRESULT CALLBACK windows_proc_callback(HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam);
+
+#define log_error(operation) do { \
+	usbi_err(NULL, "%s failed with error: %s", operation, windows_error_str(0)); \
+} while (0)
+
+int windows_start_event_monitor(void)
+{
+	windows_event_thread_handle = CreateThread(
+		NULL,
+		0,
+		windows_event_thread_main,
+		NULL,
+		0,
+		NULL
+	);
+
+	if(windows_event_thread_handle == NULL)
+	{
+		log_error("CreateThread");
+		return LIBUSB_ERROR_OTHER;
+	}
+
+	return LIBUSB_SUCCESS;
+}
+
+int windows_stop_event_monitor(void)
+{
+	if(windows_event_hWnd == NULL)
+	{
+		return LIBUSB_SUCCESS;
+	}
+
+	if(!SUCCEEDED(SendMessage(windows_event_hWnd, WM_CLOSE, 0, 0)))
+	{
+		log_error("SendMessage");
+	}
+
+	int ret = LIBUSB_SUCCESS;
+
+	if(WaitForSingleObject(windows_event_thread_handle, INFINITE) != WAIT_OBJECT_0)
+	{
+		log_error("WaitForSingleObject");
+		ret = LIBUSB_ERROR_OTHER;
+	}
+
+	if(!CloseHandle(windows_event_thread_handle))
+	{
+		log_error("WaitForSingleObject");
+		ret = LIBUSB_ERROR_OTHER;
+	}
+
+	return ret;
+}
+
+static int windows_get_device_list(struct libusb_context *ctx)
```

Comment:
nit: We should add a comment to explain what is happening here. I propose:

// We leverage the backend's get_device_list method to discover new/gone devices.
// 1. Mark all devices as is they are DISCONNECTED.
// 2. Run get_device_list which will set devices still connected to CONNECTED and new devices to NEWLY_CONNECTED.
//  Since the marker is not modified for devices which had disconnected, all devices have appropriate connection status.

Reply:
Thanks!